File: linked-list-cycle-ii/linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        slow_p = head
        fast_p = head
        start=head
        while (fast_p and fast_p.next):
            slow_p=slow_p.next
            fast_p=fast_p.next.next
            if slow_p == fast_p:
                
                while (start != slow_p):
                    start=start.next
                    slow_p=slow_p.next
                return start       
        return None
        
        """
        :type head: ListNode
        :rtype: ListNode
        """
        

File: linked-list-cycle-ii/linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
       auto slow=head;
        auto fast=head;
        auto start=head;
        while(fast and fast->next)
        {
            slow=slow->next;
            fast=fast->next->next;
            if(slow==fast)
            {
                while(start != slow)
                {
                    start=start->next;
                    slow=slow->next;
                }
                return start;
            }
        }
        return NULL;
        
    }
};